tests/ops/test_gaussian.py

```diff
@@ -4,23 +4,127 @@
+    log_normalizer = torch.randn(log_normalizer_shape)
+    info_vec = torch.randn(info_vec_shape + (dim,))
+    precision = torch.randn(info_vec_shape + (dim, rank))
+    precision = precision.matmul(precision.transpose(-1, -2))
+    gaussian = Gaussian(log_normalizer, info_vec, precision)
+
+    expected_shape = extra_shape + broadcast_shape(
+        log_normalizer_shape, info_vec_shape, precision_shape)
+    actual = gaussian.expand(expected_shape)
+    assert actual.batch_shape == expected_shape
+
+
+@pytest.mark.parametrize("old_shape,new_shape", [
+    ((6,), (3, 2)),
+    ((5, 6), (5, 3, 2)),
+], ids=str)
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_reshape(old_shape, new_shape, dim):
+    gaussian = random_gaussian(old_shape, dim)
+
+    # reshape to new
+    new = gaussian.reshape(new_shape)
+    assert new.batch_shape == new_shape
+
+    # reshape back to old
+    g = new.reshape(old_shape)
+    assert_close_gaussian(g, gaussian)
+
+
+@pytest.mark.parametrize("shape,cat_dim,split", [
+    ((4, 7, 6), -1, (2, 1, 3)),
+    ((4, 7, 6), -2, (1, 1, 2, 3)),
+    ((4, 7, 6), 1, (1, 1, 2, 3)),
+], ids=str)
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_cat(shape, cat_dim, split, dim):
+    assert sum(split) == shape[cat_dim]
+    gaussian = random_gaussian(shape, dim)
+    parts = []
+    end = 0
+    for size in split:
+        beg, end = end, end + size
+        if cat_dim == -1:
+            part = gaussian[..., beg: end]
+        elif cat_dim == -2:
+            part = gaussian[..., beg: end, :]
+        elif cat_dim == 1:
+            part = gaussian[:, beg: end]
+        else:
+            raise ValueError
+        parts.append(part)
+
+    actual = Gaussian.cat(parts, cat_dim)
+    assert_close_gaussian(actual, gaussian)
+
+
+@pytest.mark.parametrize("shape", [(), (4,), (3, 2)], ids=str)
+@pytest.mark.parametrize("dim", [1, 2, 3])
+@pytest.mark.parametrize("left", [0, 1, 2])
+@pytest.mark.parametrize("right", [0, 1, 2])
+def test_pad(shape, left, right, dim):
+    expected = random_gaussian(shape, dim)
+    padded = expected.pad(left=left, right=right)
+    assert padded.batch_shape == expected.batch_shape
+    assert padded.dim() == left + expected.dim() + right
+    mid = slice(left, padded.dim() - right)
+    assert_close(padded.info_vec[..., mid], expected.info_vec)
+    assert_close(padded.precision[..., mid, mid], expected.precision)
+
+
+@pytest.mark.parametrize("shape", [(), (4,), (3, 2)], ids=str)
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_add(shape, dim):
+    x = random_gaussian(shape, dim)
+    y = random_gaussian(shape, dim)
+    value = torch.randn(dim)
+    assert_close((x + y).log_density(value), x.log_density(value) + y.log_density(value))
+
+
+@pytest.mark.parametrize("batch_shape", [(), (4,), (3, 2)], ids=str)
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_logsumexp(batch_shape, dim):
+    gaussian = random_gaussian(batch_shape, dim)
+    gaussian.info_vec.fill_(0)  # centered
```

Comment:
I think we need to set info_vec to nonzero values, so we can detect the missing term in `logsumexp`.

Reply:
Hmm, I'm worried that the distribution will lie outside of the `[-5,5]` uniform sampling scheme below. I've weakened this line but not removed it entirely.

Reply:
Thanks! This is also related to the unstable problem of tensordot density test. I'll adopt this to make that test more stable (currently there is a single test case failing in Travis but there might be more for the next PyTorch version).